### ISSUE TITLE
bleadperl/perlio vsprintf dumps core on LinuxPPC

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2848]
Migrated from [rt.perl.org#4683](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=4683) (status was 'resolved')

Searchable as RT4683$

**Comment by p5pRT** [comment]
### From domo@computer.org

This with perl@&#8203;7638 built with \-Dperlio \(which makes it too sick to run perlbug\)&#8203;:

Many tests fail before or soon after starting by dumping core\.  The backtraces look like&#8203;:

\#0  0x1012f724 in PerlIO\_vprintf \(f=0x10164d00\, fmt=0x1014053c "%ld"\, 
&nbsp;    ap=0x7ffff730\) at perlio\.c&#8203;:2161
2161     s = SvPV\(sv\,len\);
\(gdb\) bt
\#0  0x1012f724 in PerlIO\_vprintf \(f=0x10164d00\, fmt=0x1014053c "%ld"\, 
&nbsp;    ap=0x7ffff730\) at perlio\.c&#8203;:2161
\#1  0x1012f83c in PerlIO\_printf \(f=0x10164d00\, fmt=0x1014053c "%ld"\)
&nbsp;    at perlio\.c&#8203;:2172
\#2  0x1011018c in Perl\_do\_print \(sv=0x10179c50\, fp=0x10164d00\) at doio\.c&#8203;:1192
\#3  0x10096050 in Perl\_pp\_print \(\) at pp\_hot\.c&#8203;:461
\#4  0x10092efc in Perl\_runops\_debug \(\) at run\.c&#8203;:56
\#5  0x1001752c in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:1452
\#6  0x10016f48 in perl\_run \(my\_perl=0x1015d6a8\) at perl\.c&#8203;:1372
\#7  0x10012a54 in main \(argc=2\, argv=0x7ffffa74\, env=0x7ffffa80\)
&nbsp;    at perlmain\.c&#8203;:52

That makes a compile warning about incompatible pointer types for ap on this line

 sv\_vcatpvf\(sv\, fmt\, &ap\);

of Perlio\_vprintf\(\) look highly suspicious\, but I have yet to work 
out the fix\. \(And I didn't make a copy of the exact warning before 
shutting down Linux\.  Sorry about that\.\)  Interestingly\, I don't see 
the warning on MachTen\, which sports an older gcc\, and passes all 
tests with perlio\.

Summary of my perl5 \(revision 5\.0 version 7 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.2\.17\, archname=ppc\-linux
&nbsp;    uname='linux daventry\.uunet\.lu 2\.2\.17 \#1 fri sep 8 12&#8203;:04&#8203;:50 edt 2000 ppc unknown '
&nbsp;    config\_args='\-de \-Duseperlio \-Dusedevel \-Doptimize=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing'
&nbsp;    ccversion=''\, gccversion='2\.95\.2 19991024 \(release/franzo\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.1\.3\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Well\, here's some deep black magic that fixes things on LinuxPPC&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- perl/perlio.c	Mon Nov 13 15:37:51 2000
+++ perl-perlio/perlio.c	Thu Nov 16 18:18:15 2000
@@ -2224,7 +2224,9 @@ PerlIO_vprintf(PerlIO *f, const char *fm
  SV *sv = newSVpvn("",0);
  char *s;
  STRLEN len;
- sv_vcatpvf(sv, fmt, &ap);
+ va_list apc;
+ __va_copy(apc, ap);
+ Perl_sv_vcatpvf(sv, fmt, &apc);
  s = SvPV(sv,len);
  return PerlIO_write(f,s,len);
 }


```
</details>
With that in place, perl@7674 builds and passes all tests.  (myconfig as
in the mail at the start of this thread\.\)  But I'm not proposing that
perl uses any macro starting with two
keep\-your\-fingers\-out\-of\-the\-machinery underscores\.  Since this appears
to be an undocumented facet of gcc/glibc \(see the mail thread on the gcc
list you pointed me at\) I have yet to find out whether a more offical
way of doing this exists\.  Even if it does\, it'll need \#ifdef\-ing
around\.  Sigh\.
